programs/clearing_house/src/math/lp.rs

```diff
@@ -0,0 +1,490 @@
+            .ok_or_else(math_error!())?;
+
+        let quote_asset_amount =
+            reserve_to_asset_amount(quote_asset_reserve_amount, amm.peg_multiplier)?;
+
+        let min_qaa = amm.minimum_quote_asset_trade_size;
+        let min_baa = amm.base_asset_amount_step_size;
+
+        if base_asset_amount.unsigned_abs() >= min_baa && quote_asset_amount >= min_qaa {
+            market_quote_asset_amount = quote_asset_amount;
+            market_base_asset_amount = base_asset_amount;
+        } else {
+            msg!(
+                "Warning: lp market position too small: {} {} :: {} {}",
+                base_asset_amount,
+                min_baa,
+                quote_asset_amount,
+                min_qaa
+            );
+
+            // no market position bc too small so give them negative upnl
+            // similar to closing their small position
+            unsettled_pnl = cast_to_i128(quote_asset_amount)?
+                .checked_add(1)
+                .ok_or_else(math_error!())?
+                .checked_mul(-1)
+                .ok_or_else(math_error!())?;
+        }
+    }
+
+    let metrics = LPMetrics {
+        fee_payment,
+        funding_payment,
+        base_asset_amount: market_base_asset_amount,
+        quote_asset_amount: market_quote_asset_amount,
+        unsettled_pnl,
+    };
+    msg!("lp metrics: {:#?}", metrics);
+
+    Ok(metrics)
+}
+
+pub fn calculate_swap_quote_reserve_delta(
+    amm: &AMM,
+    base_asset_amount: i128,
+) -> ClearingHouseResult<u128> {
+    let swap_direction = swap_direction_to_close_position(base_asset_amount);
+
+    let (new_quote_asset_reserve, _) = calculate_swap_output(
+        base_asset_amount.unsigned_abs(),
+        amm.base_asset_reserve,
+        swap_direction,
+        amm.sqrt_k,
+    )?;
+
+    let quote_asset_reserve_output =
+        abs_difference(new_quote_asset_reserve, amm.quote_asset_reserve)?;
+
+    Ok(quote_asset_reserve_output)
+}
+
+// settle and make tradeable
+pub fn get_lp_market_position_margin(
+    position: &MarketPosition,
+    market: &Market,
+) -> ClearingHouseResult<MarketPosition> {
+    // clone bc its only temporary
+    let mut position_clone = *position;
+    let mut market_clone = *market;
+
+    let total_lp_shares = market.amm.sqrt_k;
+    let lp_shares = position.lp_shares;
+
+    burn_lp_shares(&mut position_clone, &mut market_clone, lp_shares)?;
+
+    // worse case market position
+    // max ask: (sqrtk*1.4142 - base asset reserves) * lp share
+    // max bid: (base asset reserves - sqrtk/1.4142) * lp share
+
+    // TODO: make this a constant?
+    let sqrt_2_percision = 10_000_u128;
+    let sqrt_2 = 14142;
+
+    // worse case if all asks are filled
+    let ask_bounded_k = market
+        .amm
+        .sqrt_k
+        .checked_mul(sqrt_2)
+        .ok_or_else(math_error!())?
+        .checked_div(sqrt_2_percision)
+        .ok_or_else(math_error!())?;
+
+    let max_asks = ask_bounded_k
```

Comment:
just had a thought, should we use the spread reserves for this? (not completely sure)